Cannot run Mongoose on FreeRTOS over Linux port because of missing #include <signal.h>

- My goal is: run Mongoose webserver in a FreeRTOS task on Linux port for testing purposes
- My actions were: add "`MG_ARCH=MG_ARCH_CUSTOM`" compile definition and add `#define MG_ARCH MG_ARCH_FREERTOS` in `mongoose_config.h` file.
- My expectation was: be able to build
- The result I saw: cannot build because `signal` function, `SIGPIPE` and `SIG_IGN` are not declared.
- My question is: Is it possible to modify the preprocessor line in `mg_mgr_init` that check which MG_ARCH is in use and be able to run MG_ARCH_FREERTOS on Linux env?

# Environment

- mongoose version: 7.18
- Compiler/IDE and SDK: gcc version 12.2.0 (Debian 12.2.0-14+deb12u1)
- Target hardware/board: None
- Connectivity chip/module: None
- Target RTOS/OS (if applicable): FreeRTOS using Linux port

~~~c
void mg_mgr_init(struct mg_mgr *mgr) {
  memset(mgr, 0, sizeof(*mgr));
#if MG_ENABLE_EPOLL
  if ((mgr->epoll_fd = epoll_create1(EPOLL_CLOEXEC)) < 0)
    MG_ERROR(("epoll_create1 errno %d", errno));
#else
  mgr->epoll_fd = -1;
#endif
#if MG_ARCH == MG_ARCH_WIN32 && MG_ENABLE_WINSOCK
  // clang-format off
  { WSADATA data; WSAStartup(MAKEWORD(2, 2), &data); }
  // clang-format on
#elif MG_ENABLE_FREERTOS_TCP
  mgr->ss = FreeRTOS_CreateSocketSet();
#elif defined(__unix) || defined(__unix__) || defined(__APPLE__)
  // Ignore SIGPIPE signal, so if client cancels the request, it
  // won't kill the whole process.
  #include <signal.h>
  signal(SIGPIPE, SIG_IGN);
#elif MG_ENABLE_TCPIP_DRIVER_INIT && defined(MG_TCPIP_DRIVER_INIT)
  MG_TCPIP_DRIVER_INIT(mgr);
#endif
  mgr->pipe = MG_INVALID_SOCKET;
  mgr->dnstimeout = 3000;
  mgr->dns4.url = "udp://8.8.8.8:53";
  mgr->dns6.url = "udp://[2001:4860:4860::8888]:53";
  mg_tls_ctx_init(mgr);
}
~~~

The line does not consider `MG_ARCH_FREERTOS` or `MG_ARCH_UNIX`
~~~c
#elif defined(__unix) || defined(__unix__) || defined(__APPLE__)
~~~

and the following funcions is called but signal.h was not included in mongoose.h:
~~~c
signal(SIGPIPE, SIG_IGN);
~~~

Thanks for your reply.

**What happens when you use MG_ARCH_FREERTOS ?**
The following build error:

~~~txt
[ 13%] Building C object lib/external/CMakeFiles/mongoose.dir/mongoose/mongoose.c.o
In file included from path/to/lib/external/mongoose/mongoose.h:56,
                 from path/to/lib/external/mongoose/mongoose.c:20:
path/to/include/mongoose/mongoose_config.h:7: warning: "MG_ARCH" redefined
    7 | #define MG_ARCH MG_ARCH_FREERTOS
      | 
<command-line>: note: this is the location of the previous definition
path/to/lib/external/mongoose/mongoose.c: In function ‘mg_mgr_init’:
path/to/lib/external/mongoose/mongoose.c:4151:3: warning: implicit declaration of function ‘signal’ [-Wimplicit-function-declaration]
 4151 |   signal(SIGPIPE, SIG_IGN);
      |   ^~~~~~
path/to/lib/external/mongoose/mongoose.c:4151:10: error: ‘SIGPIPE’ undeclared (first use in this function); did you mean ‘EPIPE’?
 4151 |   signal(SIGPIPE, SIG_IGN);
      |          ^~~~~~~
      |          EPIPE
path/to/lib/external/mongoose/mongoose.c:4151:10: note: each undeclared identifier is reported only once for each function it appears in
path/to/lib/external/mongoose/mongoose.c:4151:19: error: ‘SIG_IGN’ undeclared (first use in this function)
 4151 |   signal(SIGPIPE, SIG_IGN);
~~~

**What happens when you try to define your custom arch and then go FREERTOS ? Build issue ? Runtime issue ? Which, what, where, why ?**
Build issue here:
- src: mongoose.c 
- fn: void mg_mgr_init(struct mg_mgr *mgr)
- permalink: https://github.com/cesanta/mongoose/blob/e8545858025917f14030dd0a0f0a155fc31aab92/mongoose.c#L4153

---
 I don't know if I have to consider my arch "a custom arch". I just want to simulate FreeRTOS + lwIP + Mongoose on my Linux env.

I attach the following examples and ports:
- https://github.com/FreeRTOS/FreeRTOS-Kernel/tree/main/portable/ThirdParty/GCC/Posix
- https://github.com/FreeRTOS/FreeRTOS/tree/main/FreeRTOS/Demo/Posix_GCC
- https://github.com/lwip-tcpip/lwip/tree/master/contrib/ports/unix

My CMake

~~~cmake
set(LIBRARY_NAME "mongoose")

add_library(${LIBRARY_NAME} STATIC
    ${CMAKE_CURRENT_SOURCE_DIR}/mongoose/mongoose.c
)

target_link_libraries(${LIBRARY_NAME}
        # FreeRTOSConfig.h
	freertos_config
        # FreeRTOS-Kernel
	freertos_kernel
)

target_include_directories(${LIBRARY_NAME} PUBLIC
    ${CMAKE_CURRENT_SOURCE_DIR}/mongoose
    # mongoose_config.h
    ${CMAKE_SOURCE_DIR}/include/mongoose_config
)

# Force MG_ARCH_FREERTOS inside mongoose_config.h
target_compile_definitions(${LIBRARY_NAME} PRIVATE
        MG_ARCH=MG_ARCH_CUSTOM
)
~~~

My mongoose_config.h is the following:

~~~c
#define MG_ARCH MG_ARCH_FREERTOS
#define MG_ENABLE_LWIP 1
~~~

> **What happens when you use MG_ARCH_FREERTOS ?** The following build error:
> 
> [ 13%] Building C object lib/external/CMakeFiles/mongoose.dir/mongoose/mongoose.c.o
> In file included from path/to/lib/external/mongoose/mongoose.h:56,
>                  from path/to/lib/external/mongoose/mongoose.c:20:
> path/to/include/mongoose/mongoose_config.h:7: warning: "MG_ARCH" redefined
>     7 | #define MG_ARCH MG_ARCH_FREERTOS
>       |

Can you please tell me how you are building ?
Are you explicitly passing MG_ARCH=MG_ARCH_FREERTOS as a build parameter ?
mongoose_config is intended for embedded systems, you are running under Linux, and we auto detect OS, so (I didn't actually check the code path) we may be detecting Linux and you are redefining to FreeRTOS.
Please build with `-DMG_ARCH=MG_ARCH_FREERTOS`

Using the following cmake cmd I get the same build error that you can read from my commend above:

~~~cmake
target_compile_definitions(${LIBRARY_NAME} PRIVATE
        MG_ARCH=MG_ARCH_FREERTOS
        MG_ENABLE_LWIP=1
)
~~~

I write **again** the **build error** here:

~~~txt
/home/macros/Desktop/repos/picofw/lib/external/mongoose/mongoose.c: In function ‘mg_mgr_init’:
/home/macros/Desktop/repos/picofw/lib/external/mongoose/mongoose.c:4151:3: warning: implicit declaration of function ‘signal’ [-Wimplicit-function-declaration]
 4151 |   signal(SIGPIPE, SIG_IGN);
      |   ^~~~~~
/home/macros/Desktop/repos/picofw/lib/external/mongoose/mongoose.c:4151:10: error: ‘SIGPIPE’ undeclared (first use in this function); did you mean ‘EPIPE’?
 4151 |   signal(SIGPIPE, SIG_IGN);
      |          ^~~~~~~
      |          EPIPE
/home/macros/Desktop/repos/picofw/lib/external/mongoose/mongoose.c:4151:10: note: each undeclared identifier is reported only once for each function it appears in
/home/macros/Desktop/repos/picofw/lib/external/mongoose/mongoose.c:4151:19: error: ‘SIG_IGN’ undeclared (first use in this function)
 4151 |   signal(SIGPIPE, SIG_IGN);
~~~

--- a/mongoose.c
+++ b/mongoose.c
@@ -4147,7 +4147,7 @@ void mg_mgr_init(struct mg_mgr *mgr) {
   // clang-format on
 #elif MG_ENABLE_FREERTOS_TCP
   mgr->ss = FreeRTOS_CreateSocketSet();
-#elif defined(__unix) || defined(__unix__) || defined(__APPLE__)
+#elif MG_ARCH == MG_ARCH_UNIX
   // Ignore SIGPIPE signal, so if client cancels the request, it
   // won't kill the whole process.
   signal(SIGPIPE, SIG_IGN);

--- a/src/net.c
+++ b/src/net.c
@@ -267,7 +267,7 @@ void mg_mgr_init(struct mg_mgr *mgr) {
   // clang-format on
 #elif MG_ENABLE_FREERTOS_TCP
   mgr->ss = FreeRTOS_CreateSocketSet();
-#elif defined(__unix) || defined(__unix__) || defined(__APPLE__)
+#elif MG_ARCH == MG_ARCH_UNIX
   // Ignore SIGPIPE signal, so if client cancels the request, it
   // won't kill the whole process.
   signal(SIGPIPE, SIG_IGN);

